### ISSUE TITLE
[FEATURE REQUEST] Add separate pagination limits for widgets and addon

### ISSUE THREAD

**Issue by ivarbrandt** [Tikipeter/repository.tikipeter#179]
Could you please add the ability to set separate page limits for widgets as well as the addon?

This would be extremely useful because I have the 'Next Page' item limit in widgets set to 20. I however have hundreds of movies and shows in my Trakt collection and I don't really wanna load the whole enchilada every time I hit 'Next Page' from the widgets.

**Comment by Tikipeter** [comment]
There is a separate pagination limit for widgets and addon already.

The issue you're facing is when you hit Next Page on the widget, that will enter you into the addon, so the limit of course no longer applies. Unfortunately there is not anything I can do for that. The check to see if the list items are a widget will come back false once you enter the addon.

If you turn off "Widgets Only" in the Pagination settings, then it might work the way you want.

**Comment by ivarbrandt** [comment]
> If you turn off "Widgets Only" in the Pagination settings, then it might work the way you want.

I did this and it paginates the addon correctly but I lose pagination for the widgets.

If I turn on 'Widgets Only", the widgets will be paginated but the addon won't be paginated.

> The check to see if the list items are a widget will come back false once you enter the addon.

Is there a way to implement that if the check comes back false, paginate to x number of items?

**Comment by Tikipeter** [comment]
>I did this and it paginates the addon correctly but I lose pagination for the widgets.

It isn't meant to do this. Both the widgets and the addon should limit to the page limit when "Widgets Only" isn't set.

I've just made a few widgets to test, and it looks like the IMDb widgets aren't behaving how they should. And are instead behaving like you describe. Is it an IMDb list widget that isn't limiting to the defined page limit?

**Comment by Tikipeter** [comment]
Unfortunately IMDb lists can't be paginated.

**Comment by ivarbrandt** [comment]
No it is not an IMDb list. It is a Trakt list. Specifically TRAKT movie and show collection and movie and show watchlist.

**Comment by Tikipeter** [comment]
All the Trakt lists paginate correctly according to the settings. If "Widget Only" is OFF, then I get a paginated list both within the addon and with the widget.

Are you refreshing the widgets after changing the setting? They won't refresh by themselves just because a setting has been changed.

**Comment by ivarbrandt** [comment]
> If "Widget Only" is OFF, then I get a paginated list both within the addon and with the widget

Correct. But the limit of number of items is the same for both widget and addon. 

Is is possible to make 2 different limits, one for the widgets, one for the addon?

You mentioned that there is a check to see if it is a widget. So if that check comes back false could you make it paginate the list in the addons to x number of items? This would be preferable to loading the whole list.

**Comment by Tikipeter** [comment]
>Correct. But the limit of number of items is the same for both widget and addon.

>This would be preferable to loading the whole list.

I've become confused. If you turn OFF "Widgets Only" in the settings, does both the addon and widget respect the pagination limit? If not, what widget/addon container is not honoring that setting?

Assuming the "Widgets Only" ON/OFF is working correctly, then your request is to have the widgets and addon display a different number of items per page. I don't see how this would be possible.

Say you were to set widgets to 20 items per page, and the addon to 50 items per page. When you hit the "Next Page" on the widget, you will expect the first item within the addon to be item 21 in the list. But because you have the addon set to display 50 items per page, the second page within the addon will start with item 51. You will not have access to items 21-50.

**Comment by ivarbrandt** [comment]
> If you turn OFF "Widgets Only" in the settings, does both the addon and widget respect the pagination limit?

Yes the addon and the widget both respect the pagination limit.

> Then your request is to have the widgets and addon display a different number of items per page.

Yes this is correct.

> Say you were to set widgets to 20 items per page, and the addon to 50 items per page. When you hit the "Next Page" on the widget, you will expect the first item within the addon to be item 21 in the list. But because you have the addon set to display 50 items per page, the second page within the addon will start with item 51. You will not have access to items 21-50.

Shit. This makes sense. I guess I thought that the addon could know it was a widget so only display 20 items and the 'Next Page' item would be linked to the next page of items in the addon. So it would know to load the next 50 items, starting with the 21st item in the list. That way you would be able to see items 21-71 on that page in the addon.

I guess maybe the only way to do this would be that the addon would have to know what the last item in the widget was so it knows where to start the next page of items in the addon.

**Comment by Tikipeter** [comment]
Possible, but not at all easy with the way the code is written now.

I'll leave this open and have a good think about it....

**Comment by ivarbrandt** [comment]
> When I say it would not be easy, I mean respecting the end point of the widget list when then opening the addon with the "Next Page" from the widget

I understand. I was thinking deeper into this after my last comment and I saw how it might be a pain in the ass to do this lol.

> I'll leave this open and have a good think about it....

Sounds good. Thank you for considering it.

**Comment by Tikipeter** [comment]
I've hacked something together for this for next update.

Keep in mind it will only work with Fen settings, not artificial skin widget limit settings.

**Comment by ivarbrandt** [comment]
Awesome! Can't wait to test it out. Thanks for giving it a shot.

> Keep in mind it will only work with Fen settings, not artificial skin widget limit settings.

Yeah that makes sense. I added adjustable widget limits yesterday to FENtastic so this should work quite well with it.

**Comment by Tikipeter** [comment]
No, it won't work properly if it is the skin setting the widget limit and not the addon, unless the 2 settings values match.

Widget item limits by skins is almost useless really.

If you have a widget that has 200 list items in it, then Fen (any addon) will still process and make those 200 list items. Then the skin will limit how many to display. But the time will still be taken to make all 200 list items first.

**Comment by Tikipeter** [comment]
3.1.05 is released and includes this new behavior.

**Comment by ivarbrandt** [comment]
Ahhh I see. That explains the varying loading times I've seen when loading different size widget lists set by Fen while keeping a static limit set by skins.

> Widget item limits by skins is almost useless really.

I think they will be even more useless with this new feature lol. Can you think of any use case where the skin widget limits would be useful? If not, I'll just plan on removing them on the next update.

Btw this new feature is absolute 🔥! Thanks for adding it.

**Comment by ivarbrandt** [comment]
> I suppose if you aren't limiting the widget so as to reduce loading times

That's exactly why I had skin limits set. I added adjustable limits to my skin to give more control over how many items are displayed per widget. Now I know that is not related to widget loading times.

That being said, I'm going to leave the option to adjust the limits through the skin because I can see use cases where people use a mix of a native Kodi library and addons.

Thanks for all the info. I always learn something new when I talk to you.
